File: data_structures/heap.js
function Heap(){
    this.store = []
}

Heap.parentIndex = function(childIdx){
    if(childIdx === 0){
        return;
    }else{
        return Math.floor((childIdx -1)/2);
    }
}

Heap.childIndices = function(len, parentIdx){
    var indices = [];
    var potential_childs = [(2*parentIdx)+1, (2*parentIdx)+2]
    potential_childs.forEach(child => {
        if(child < len){
            indices.push(child);
        }
    })

    return indices;
}

Heap.heapifyUp = function(arr, childIdx, len = arr.length, comp){
    if(!comp){
        comp = function(a, b){
            return a-b;
        }
    }

    var swapped = true;
    while(swapped){
        swapped = false;
        var parentIdx = Heap.parentIndex(childIdx);
        if(parentIdx !== undefined){
            if(comp(arr[childIdx], arr[parentIdx]) < 0){
                var temp = arr[childIdx];
                arr[childIdx] = arr[parentIdx];
                arr[parentIdx] = temp;
                swapped = true;
                childIdx = parentIdx;
            }
        }
    }
    return arr;
}

Heap.heapifyDown = function(arr, parentIdx, len = arr.length, comp){
    if(!comp) {
        comp = function(a,b){
            return a-b;
        }
    }

    var swapped = true;
    while(swapped){
        swapped = false;
        var minIndex;
        var childIndices = Heap.childIndices(len, parentIdx);
        if(childIndices.length > 1){
            var compVal = comp(arr[childIndices[0]], arr[childIndices[1]]);
            if(compVal > 0){
                minIndex = childIndices[1];
            }else{
                minIndex = childIndices[0];
            }
        }else if(childIndices.length === 1){
            minIndex = childIndices[0];
        }else{
            return arr;
        }

        if(comp(arr[parentIdx], arr[minIndex]) > 0){
            var temp = arr[parentIdx];
            arr[parentIdx] = arr[minIndex];
            arr[minIndex] = temp;
            swapped = true;
            parentIdx = minIndex;
        }
    }
    return arr;
}

Heap.prototype.push = function(val){
    this.store.push(val);
    var childIdx = this.store.length -1;
    Heap.heapifyUp(this.store, childIdx);
}

Heap.prototype.peek = function(){
    return this.store[0];
}

Heap.prototype.extract = function(){
    var temp = this.store[0];
    this.store[0] = this.store[this.store.length-1];
    this.store[this.store.length-1] = temp;
    var el = this.store.pop();
    Heap.heapifyDown(this.store, 0);
    return el;

}

File: data_structures/heaps/lib/heap.rb
class BinaryMinHeap
  def initialize(&prc)
    @store = []
  end

  def count
    @store.length
  end

  def extract
    @store[0], @store[-1] = @store[-1], @store[0]
    el = @store.pop
    BinaryMinHeap.heapify_down(@store, 0)
    el
  end

  def peek
    @store[0]
  end

  def push(val)
    @store.push(val)
    index = self.count - 1
    BinaryMinHeap.heapify_up(@store, index)
  end

  protected
  attr_accessor :prc, :store

  public
  def self.child_indices(len, parent_index)
    indices = []
    potential_childs = [(2*parent_index)+1, (2*parent_index)+2]
    potential_childs.each {|index| indices.push(index) if index < len}
    indices
  end

  def self.parent_index(child_index)
    if child_index == 0
      raise "root has no parent"
    else
      (child_index-1)/2
    end
  end

  def self.heapify_down(array, parent_idx, len = array.length, &prc)
    if !prc
      prc = Proc.new {|a, b| a <=> b}
    end

    swapped = true
    while swapped
      swapped = false
      el = array[parent_idx]
      child_indices = BinaryMinHeap.child_indices(len, parent_idx)
      min = 0
      if child_indices.length > 1
        comparator = prc.call(array[child_indices[0]], array[child_indices[1]])
        if comparator > 0
          min = child_indices[1]
        else
          min = child_indices[0]
        end
      elsif child_indices.length == 1
        min = child_indices[0]
      else
        return array
      end

      if prc.call(array[parent_idx], array[min]) > 0
        array[parent_idx], array[min] = array[min], array[parent_idx]
        swapped = true
        parent_idx = min
      end
    end
    array
  end

  def self.heapify_up(array, child_idx, len = array.length, &prc)
    if !prc
      prc = Proc.new {|a, b| a <=> b}
    end

    swapped = true
    while swapped
      swapped = false
      el = array[child_idx]
      if child_idx > 0
        parent_idx = BinaryMinHeap.parent_index(child_idx)
      else
        return array
      end
      if prc.call(el, array[parent_idx]) < 0
        array[child_idx], array[parent_idx] = array[parent_idx], array[child_idx]
        child_idx = parent_idx
        swapped = true
      end
    end

    array
  end
end
